### ISSUE TITLE
Option to remove user from single message 

### ISSUE THREAD

**Issue by j-m-nunes** [sympa-community/sympa#1531]
Add an option to exclude some recipients from receiving a message

Expected Behavior
-----------------
For instance
to: mylist@mail1 except: user1@mail1, user2@mail2
would deliver the message to the list without sending it to the users mentioned in except (or exclude or ...)

Current Behavior
----------------
The behavior does not exist and that causes people to send message to all recipients in cleat to: or cc: exposing often publicly lots of addresses

Context
-------
This would be useful for instance when a party is being organized to some one that leaves or graduates or is being promoted, when everyone except a few persons in an already existing list should receive a message.
This would improve security and reduce phishing-like address recollection.

**Comment by ikedas** [comment]
You should use BCC instead of To/CC so that your concern about security risk will be solved.

**Comment by ikedas** [comment]
Currently Sympa is not specialized for distributing newsletters, so dynamic distribution list generation based on subscriber attributes is not possible. Software that can do that includes [phplist](https://www.phplist.org/) (along with [segment plugin](https://resources.phplist.com/plugin/segment) or [attribute select plugin](https://resources.phplist.com/plugin/attributeselect)).

**Comment by j-m-nunes** [comment]
Thanks for the suggestions but BCC: defeats the idea that everyone sees who received the message and who did not. Suppose it is a message to discuss a gift for someone who graduated, that person should not be aware of the messages, all others should.
I don't know the phpList but the mailing software used in my organization is sympa so it would be nice to have the feature added to it.
Off course such a message based feature may be difficult to add. I thought it would not be very different from controlling for multiple copies to the same recipient.

**Comment by racke** [comment]
You could set reception for the excluded users to nomail, send the message and set the reception back to the previous value. But you need to be aware that the message will show up in the archives if they are enabled.

**Comment by j-m-nunes** [comment]
Thanks. If I understand correctly your suggestion it implies admin access to the list and the idea was to make the option available directly for all users.

**Comment by racke** [comment]
Yes, that implies admin access to the list. So how do you imagine selecting the user and sending the message? Both in the web interface?

**Comment by j-m-nunes** [comment]
I was thinking of a custom mail header, say exclude:, those in these header will not receive the message. Adding this custom mail header to the mail clients can easily be done by the final user (or deployed by the organization).

**Comment by ikedas** [comment]
I don't think there are email clients that the average user uses today that allows to easily add arbitrary header fields.

To begin with, if you know who the users are that you want to exclude, then you have a list of all the subscribers' addresses on the mailing list. Then, you could just send BCC with only the addresses of the users you want to send a message to.

**Comment by j-m-nunes** [comment]
Custom header for Thunderbird is easy to do :
https://www.lifewire.com/arbitrary-custom-heading-email-thunderbird-1173089

Such an exclude feature is really useful and an improvement if any list user can use it. For controlled or centralised situations your suggestions are a cleaner way of doing it. Just in my university I guesstimate between one and ten thousand messages a month would benefit from it.

**Comment by racke** [comment]
It might be technical possible to implement this feature with a plugin using the message hook of Sympa.

**Comment by ldidry** [comment]
> Currently Sympa is not specialized for distributing newsletters, so dynamic distribution list generation based on subscriber attributes is not possible. Software that can do that includes [phpList](https://www.phplist.org/) (along with [segment plugin](https://resources.phplist.com/plugin/segment) or [attribute select plugin](https://resources.phplist.com/plugin/attributeselect)).

You may have a look at [Listmonk](https://github.com/knadh/listmonk/).

**Comment by j-m-nunes** [comment]
I do understand your suggestions but I think my proposal is not for a modified newsletter but a legitimate use of a list as the messages can be originated and replied by any list member. At any given point in time there may be messages excluding some user and other messages excluding other users while most of the messages in the list are still received by all list members.
As I don't know phpList I'll give it a look and if it seems a solution to my suggestion I'll acknowledge it here and close the issue.

**Comment by ikedas** [comment]
> Custom header for Thunderbird is easy to do : https://www.lifewire.com/arbitrary-custom-heading-email-thunderbird-1173089

This is a software setting, and it seems not intended to be changed every time email is sent. In fact, If a user sends an email to a list adding an exclude header with this setting, there is a big risk that they will later forget about it and send the email to the list and not send it to everyone they expected to. 

> Such an exclude feature is really useful and an improvement if any list user can use it. For controlled or centralised situations your suggestions are a cleaner way of doing it. Just in my university I guesstimate between one and ten thousand messages a month would benefit from it.

I don't think phpList fits your requirements. However, the other method I suggested, BCC, is very flexible and probably fits your requirements well.

In the first place, the point I can't get around is that, if you have to decide who to exclude each time, i.e. who to send each time, you don't seem to need to use the mailing list.

**Comment by ikedas** [comment]
One feature of Sympa that is somewhat similar to @j-m-nunes's proposal is the **message topics**.
https://www.sympa.community/manual/customize/basics-delivery.html#message-topics

Subscribers of the list may "subscribe" to any of several "topics" that the list owner has prepared in advance. If a keyword related to a topic is found in the subject or body of a message sent by the sender, the message will be received only by subscribers who subscribe to that topic. (If no keywords related to any topic are found, the message will be received by all subscribers. Subscribers who are not subscribed to any topic will receive all messages.)

This feature does not provide complete control over who is excluded from receiving messages, but it can reduce the number of messages each subscriber receives on topics they are not interested in.
